app/routines/find_or_create_unclaimed_user.rb

```diff
@@ -0,0 +1,90 @@
+# Find or create a new user with state "unclaimed"
+#
+# Given either an email address or username:
+#   attempt to find a user with that attribute.
+#   If the user is found, return the user
+#     Otherwise create a new user with the email and username given,
+#     set it's state to "unclaimed" and return that record
+
+class FindOrCreateUnclaimedUser
+
+  lev_routine
+
+  uses_routine CreateUser, translations: { outputs: { type: :verbatim } }
+  uses_routine AddEmailToUser
+  uses_routine SetPassword
+  uses_routine CreateIdentity
+
+  protected
+
+  def exec(options)
+    user = nil
+
+    if options[:username]
+      user = find_or_create_by_username(options)
+    elsif options[:email]
+      user = find_or_create_by_email(options)
+    else
+      fatal_error(code: :invalid_input, message: "Must provide either email or username")
+      return
+    end
+
+    if 'unclaimed' == user.state
+      # If a username and password was given, set the unclaimed user's identity
+      if options[:username] && options[:password]
+        set_or_create_password(user, options)
+      end
+      outputs[:user] = user
+    else
+      fatal_error(code: :account_already_claimed, message: "Account has already been claimed")
+      return
+    end
+  end
+
+
+  def set_or_create_password(user, options)
+    if user.identity
```

Comment:
If a user already has an identity, is this code resetting their password?  Should an application really have a right to do that?  I tend to think not.  But if an application sends password information, we'd need a way to tell the application that the password was not changed.